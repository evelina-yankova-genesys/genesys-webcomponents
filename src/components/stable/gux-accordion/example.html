<h1>gux-accordion</h1>

<gux-accordion arrow-position="beside-text">
  <gux-accordion-section disabled>
    <h2 slot="header">First Section Beta</h2>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">Second Section Beta</h2>
    <p slot="content">I'm a p.</p>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">Third Section Beta</h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">
      Fourth Section Beta has a really really long title to see what it looks
      like when the title overflows
    </h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
</gux-accordion>

<h2>Single Open Section</h2>

<gux-accordion arrow-position="beside-text" single-open-section>
  <gux-accordion-section>
    <h2 slot="header">First Section Beta</h2>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">Second Section Beta</h2>
    <p slot="content">I'm a p.</p>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">Third Section Beta</h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">
      Fourth Section Beta has a really really long title to see what it looks
      like when the title overflows
    </h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
</gux-accordion>

<h2>Beside Text Arrow Positioning</h2>

<gux-accordion arrow-position="beside-text">
  <gux-accordion-section>
    <h2 slot="header">First Section Beta</h2>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">Second Section Beta</h2>
    <p slot="content">I'm a p.</p>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">Third Section Beta</h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
  <gux-accordion-section>
    <h2 slot="header">
      Fourth Section Beta has a really really long title to see what it looks
      like when the title overflows
    </h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
</gux-accordion>

<h2>Invalid Headers</h2>

<gux-accordion>
  <gux-accordion-section arrow-position="beside-text">
    <div slot="header">First Section Beta</div>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
  <gux-accordion-section arrow-position="beside-text">
    <div slot="header">Second Section Beta</div>
    <p slot="content">I'm a p.</p>
  </gux-accordion-section>
  <gux-accordion-section arrow-position="beside-text">
    <div slot="header">Third Section Beta</div>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
  <gux-accordion-section arrow-position="beside-text">
    <div slot="header">
      Fourth Section Beta has a really really long title to see what it looks
      like when the title overflows
    </div>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
</gux-accordion>

<h2>All sections initially open</h2>

<gux-accordion arrow-position="beside-text">
  <gux-accordion-section class="blocked" disabled open>
    <h2 slot="header">First Section Beta</h2>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
  <gux-accordion-section open>
    <h2 slot="header">Second Section Beta</h2>
    <p slot="content">I'm a p.</p>
  </gux-accordion-section>
  <gux-accordion-section open>
    <h2 slot="header">Third Section Beta</h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
  <gux-accordion-section open>
    <h2 slot="header">
      Fourth Section Beta has a really really long title to see what it looks
      like when the title overflows
    </h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
</gux-accordion>
<h2>All sections initially open but single section open set</h2>

<gux-accordion arrow-position="beside-text" single-open-section>
  <gux-accordion-section class="blocked" disabled open>
    <h2 slot="header">First Section Beta</h2>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
  <gux-accordion-section open>
    <h2 slot="header">Second Section Beta</h2>
    <p slot="content">I'm a p.</p>
  </gux-accordion-section>
  <gux-accordion-section open>
    <h2 slot="header">Third Section Beta</h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
  <gux-accordion-section open>
    <h2 slot="header">
      Fourth Section Beta has a really really long title to see what it looks
      like when the title overflows
    </h2>
    <span slot="content">I'm a span.</span>
  </gux-accordion-section>
</gux-accordion>

<h2>With icons & subheadings</h2>
<gux-accordion>
  <gux-accordion-section>
    <h2 slot="header">Section with subheading</h2>
    <h3 slot="subheading">This is the subheading</h3>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>

  <gux-accordion-section reverse-headings="true">
    <h2 slot="header">Section with heading position reversed</h2>
    <h3 slot="subheading">This is the subheading</h3>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>

  <gux-accordion-section>
    <h2 slot="header">Section with icon</h2>
    <gux-icon
      slot="icon"
      icon-name="add-user"
      screenreader-text="add John Smith to contact list"
    ></gux-icon>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>

  <gux-accordion-section>
    <h2 slot="header">Section with icon</h2>
    <gux-icon
      slot="icon"
      icon-name="add-user"
      screenreader-text="add John Smith to contact list"
    ></gux-icon>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>

  <gux-accordion-section>
    <h2 slot="header">Section with icon & subheading</h2>
    <h3 slot="subheading">This is the subheading</h3>
    <gux-icon
      slot="icon"
      icon-name="add-user"
      screenreader-text="add John Smith to contact list"
    ></gux-icon>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>

  <gux-accordion-section reverse-headings="true">
    <h2 slot="header">Section with icon & heading position reversed</h2>
    <h3 slot="subheading">This is the subheading</h3>
    <gux-icon
      slot="icon"
      icon-name="add-user"
      screenreader-text="add John Smith to contact list"
    ></gux-icon>
    <div slot="content">
      <span>I'm a span in a div.</span>
      <button>I'm the button.</button>
    </div>
  </gux-accordion-section>
</gux-accordion>
