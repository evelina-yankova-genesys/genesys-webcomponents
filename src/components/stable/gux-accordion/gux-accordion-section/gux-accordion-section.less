@import (reference) '../../../../style/color.less';
@import (reference) '../../../../style/mixins.less';
@import (reference) '../../../../style/spacing.less';
@import (reference) '../../../../style/typography.less';

:host {
  color: @gux-black-50;
  -webkit-user-select: none;
  user-select: none;
  background-color: @gux-grey-100;

  &:first-child {
    .gux-header {
      border-top: 1px solid @gux-grey-60;
    }
  }
}

section.gux-disabled {
  cursor: default;
  opacity: 0.5;

  > * {
    pointer-events: none;
  }
}

.gux-header {
  all: unset;
  box-sizing: border-box;
  display: flex;
  flex-direction: row;
  flex-wrap: nowrap;
  align-content: center;
  align-items: center;
  justify-content: flex-start;
  width: 100%;
  padding: var(--spacing-small) var(--spacing-small) var(--spacing-small)
    var(--spacing-medium);
  margin: 0;
  cursor: pointer;
  border-bottom: 1px solid @gux-grey-60;

  .gux-header-text {
    flex: 1 1 auto;
    align-self: auto;
    padding-right: @spacing-xs;
    text-align: left;

    &.arrow-position-beside {
      flex-grow: 0;
    }
  }

  &.reverse-headings {
    & .gux-header-text {
      display: flex;
      flex-direction: column-reverse;
    }

    & svg {
      color: red !important;
    }
  }

  ::slotted([slot='header']) {
    padding: 0;
    margin: 0;
    font-size: var(--font-size-200) !important;
    line-height: var(--line-height-100) !important;
  }

  ::slotted([slot='subheading']) {
    padding: 0;
    margin: 2px 0 0;
    font-size: var(--font-size-100) !important;
    font-weight: var(--font-weight-normal) !important;
  }

  ::slotted([slot='icon']) {
    align-self: flex-start;
    width: 16px;
    height: 16px;
    margin-right: 8px;
    color: var(--gux-black-100);

    .reverse-headings& {
      align-self: flex-end;
    }
  }

  .gux-header-icon {
    flex: 0 1 auto;
    align-items: center;
    align-self: auto;
    margin-left: var(--spacing-xs);
    color: @gux-black-100;
    transition: transform 0.5s ease;
    transform-origin: center;

    gux-icon {
      width: var(--spacing-medium);
      height: var(--spacing-medium);
    }

    &.gux-expanded {
      transform: rotate(-180deg);
    }
  }

  &:focus {
    .focus-ring();

    .gux-header-icon {
      color: @gux-black-50;
    }
  }

  &:hover {
    .gux-header-icon {
      color: @gux-black-50;
    }
  }
}

.gux-content {
  box-sizing: border-box;
  display: none;
  padding: @spacing-medium;
  border-bottom: 1px solid @gux-grey-60;

  &.gux-expanded {
    display: block;
  }
}
