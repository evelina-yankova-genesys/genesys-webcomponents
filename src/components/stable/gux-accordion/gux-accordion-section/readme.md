# gux-accordion



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute          | Description | Type                         | Default     |
| ----------------- | ------------------ | ----------- | ---------------------------- | ----------- |
| `arrowPosition`   | `arrow-position`   |             | `"beside-text" \| "default"` | `'default'` |
| `disabled`        | `disabled`         |             | `boolean`                    | `false`     |
| `open`            | `open`             |             | `boolean`                    | `false`     |
| `reverseHeadings` | `reverse-headings` |             | `boolean`                    | `false`     |


## Events

| Event                   | Description | Type                |
| ----------------------- | ----------- | ------------------- |
| `internalsectionopened` |             | `CustomEvent<void>` |


## Dependencies

### Depends on

- [gux-icon](../../gux-icon)

### Graph
```mermaid
graph TD;
  gux-accordion-section --> gux-icon
  style gux-accordion-section fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
