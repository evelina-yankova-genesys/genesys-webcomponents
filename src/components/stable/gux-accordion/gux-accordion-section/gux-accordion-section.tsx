import {
  Component,
  Element,
  Event,
  EventEmitter,
  h,
  JSX,
  Prop,
  Watch
} from '@stencil/core';

import { randomHTMLId } from '@utils/dom/random-html-id';
import { logError } from '@utils/error/log-error';

import { GuxAccordionSectionArrowPosition } from './gux-accordion-section.types';

@Component({
  styleUrl: 'gux-accordion-section.less',
  tag: 'gux-accordion-section',
  shadow: true
})
export class GuxAccordionSection {
  private sectionId: string = randomHTMLId('gux-accordion-section');

  @Element()
  root: HTMLElement;

  @Prop()
  arrowPosition: GuxAccordionSectionArrowPosition = 'default';

  @Prop({ mutable: true })
  open: boolean = false;

  @Prop()
  disabled: boolean = false;

  @Prop()
  reverseHeadings: boolean = false;

  @Event()
  internalsectionopened: EventEmitter<void>;

  @Watch('open')
  watchOpen(open: boolean): void {
    if (open) {
      this.internalsectionopened.emit();
    }
  }

  private toggle() {
    this.open = !this.open;
  }

  private isArrowPositionedBesideText(): boolean {
    if (this.arrowPosition === 'beside-text') {
      return true;
    }
    return false;
  }

  private handleSlotChange(slotname: string): void {
    const slot = this.root.querySelector(`[slot="${slotname}"]`);

    if (!slot || !/^H[1-6]$/.test(slot.nodeName)) {
      logError(
        'gux-accordion-section',
        `For accessibility reasons the ${slotname} slot should be filled with a HTML heading tag (h1 - h6).`
      );
    }
  }

  render(): JSX.Element {
    return (
      <section class={{ 'gux-disabled': this.disabled }}>
        <button
          class={{
            'gux-header': true,
            'reverse-headings': this.reverseHeadings
          }}
          aria-expanded={this.open.toString()}
          aria-controls={this.sectionId}
          disabled={this.disabled}
          onClick={this.toggle.bind(this)}
        >
          <slot name="icon"></slot>
          <div
            class={{
              'gux-header-text': true,
              'arrow-position-beside': this.isArrowPositionedBesideText()
            }}
          >
            <slot
              onSlotchange={() => this.handleSlotChange('header')}
              name="header"
            ></slot>
            <slot
              onSlotchange={() => this.handleSlotChange('subheading')}
              name="subheading"
            ></slot>
          </div>
          <div
            class={{
              'gux-header-icon': true,
              'gux-expanded': this.open
            }}
          >
            <gux-icon decorative icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
        <div
          id={this.sectionId}
          class={{
            'gux-content': true,
            'gux-expanded': this.open
          }}
        >
          <slot name="content"></slot>
        </div>
      </section>
    ) as JSX.Element;
  }
}
