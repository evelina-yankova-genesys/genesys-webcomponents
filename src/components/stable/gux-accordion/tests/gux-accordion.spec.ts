import { newSpecPage } from '@stencil/core/testing';
import { GuxAccordion } from '../gux-accordion';
import { GuxAccordionSection } from '../gux-accordion-section/gux-accordion-section';
import { html } from './gux-accordion.markup';

const components = [GuxAccordion, GuxAccordionSection];
// const html = `
// <gux-accordion>
//   <gux-accordion-section>
//     <h2 slot="header">First Section</h2>
//     <div slot="content">Sample Content</div>
//   </gux-accordion-section>
//   <gux-accordion-section arrow-position="beside-text">
//     <h2 slot="header">Second Section</h2>
//     <div slot="content">Sample Content</div>
//   </gux-accordion-section>
//   <gux-accordion-section disabled>
//     <h2 slot="header">Third Section</h2>
//     <h3 slot="subheading">This is the subheading</h3>
//     <div slot="content">Sample Content</div>
//   </gux-accordion-section>
//   <gux-accordion-section>
//     <h2 slot="header">Fourth Section</h2>
//     <h3 slot="subheading">This is the subheading</h3>
//     <gux-icon
//       slot="icon"
//       icon-name="add-user"
//       screenreader-text="add John Smith to contact list"
//     ></gux-icon>
//     <div slot="content">Sample Content</div>
//   </gux-accordion-section>
//   <gux-accordion-section reverse-headings="true">
//     <h2 slot="header">Fourth Section</h2>
//     <h3 slot="subheading">This is the subheading</h3>
//     <gux-icon
//       slot="icon"
//       icon-name="add-user"
//       screenreader-text="add John Smith to contact list"
//     ></gux-icon>
//     <div slot="content">Sample Content</div>
//   </gux-accordion-section>
// </gux-accordion>
// `;
const language = 'en';

describe('gux-accordion', () => {
  it('should build', async () => {
    const page = await newSpecPage({ components, html, language });

    expect(page.rootInstance).toBeInstanceOf(GuxAccordion);
  });

  it('renders', async () => {
    const page = await newSpecPage({ components, html, language });

    expect(page.root).toMatchSnapshot();
  });
});
