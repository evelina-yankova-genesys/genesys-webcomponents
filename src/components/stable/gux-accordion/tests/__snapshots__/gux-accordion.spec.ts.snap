// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-accordion renders 1`] = `
<gux-accordion>
  <mock:shadow-root>
    <slot></slot>
  </mock:shadow-root>
  <gux-accordion-section>
    <mock:shadow-root>
      <section>
        <button aria-controls="gux-accordion-section-i" aria-expanded="false" class="gux-header">
          <slot name="icon"></slot>
          <div class="gux-header-text">
            <slot name="header"></slot>
            <slot name="subheading"></slot>
          </div>
          <div class="gux-header-icon">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
        <div class="gux-content" id="gux-accordion-section-i">
          <slot name="content"></slot>
        </div>
      </section>
    </mock:shadow-root>
    <h2 slot="header">
      First Section
    </h2>
    <div slot="content">
      Sample Content
    </div>
  </gux-accordion-section>
  <gux-accordion-section>
    <mock:shadow-root>
      <section>
        <button aria-controls="gux-accordion-section-i" aria-expanded="false" class="gux-header">
          <slot name="icon"></slot>
          <div class="gux-header-text">
            <slot name="header"></slot>
            <slot name="subheading"></slot>
          </div>
          <div class="gux-header-icon">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
        <div class="gux-content" id="gux-accordion-section-i">
          <slot name="content"></slot>
        </div>
      </section>
    </mock:shadow-root>
    <h2 slot="header">
      Second Section
    </h2>
    <div slot="content">
      Sample Content
    </div>
  </gux-accordion-section>
  <gux-accordion-section>
    <mock:shadow-root>
      <section>
        <button aria-controls="gux-accordion-section-i" aria-expanded="false" class="gux-header">
          <slot name="icon"></slot>
          <div class="gux-header-text">
            <slot name="header"></slot>
            <slot name="subheading"></slot>
          </div>
          <div class="gux-header-icon">
            <gux-icon decorative="" icon-name="chevron-small-down"></gux-icon>
          </div>
        </button>
        <div class="gux-content" id="gux-accordion-section-i">
          <slot name="content"></slot>
        </div>
      </section>
    </mock:shadow-root>
    <h2 slot="header">
      Third Section
    </h2>
    <div slot="content">
      Sample Content
    </div>
  </gux-accordion-section>
</gux-accordion>
`;
